test_util/azure.py

```diff
@@ -0,0 +1,170 @@
+        template = TemplateLink(uri=template_uri)
+        azure_wrapper.rmc.resource_groups.create_or_update(
+            group_name,
+            ResourceGroup(location=location))
+        template_params = {
+            'sshRSAPublicKey': public_key,
+            # must be lower or validation will fail
+            'masterEndpointDNSNamePrefix': master_prefix,
+            'agentEndpointDNSNamePrefix': agent_prefix,
+            'linuxAdminUsername': admin_name,
+            'agentVMSize': vm_size,
+            'agentCount': agent_count,
+            'nameSuffix': name_suffix,
+            'oauthEnabled': repr(oauth_enabled).lower(),
+            # oauth uses string, vm diagnostic uses bool
+            'enableVMDiagnostics': vm_diagnostics_enabled}
+        log.info('Provided template parameters: {}'.format(template_params))
+        # add worthless required 'value' fields
+        template_parameters = {k: {'value': v} for k, v in template_params.items()}
+        deployment_properties = DeploymentProperties(
+            template_link=template,
+            mode=DeploymentMode.incremental,
+            parameters=template_parameters)
+        # Use RPC against azure to validate the ARM template is well-formed
+        result = azure_wrapper.rmc.deployments.validate(
+            group_name, deployment_name, properties=deployment_properties)
+        if result.error:
+            for details in result.error.details:
+                log.error('{}: {}'.format(details.code, details.message))
+            raise Exception("Template verification failed!")
+
+        log.info('Starting template deployment')
+        deploy_poller = azure_wrapper.rmc.deployments.create_or_update(
+            group_name, deployment_name, deployment_properties)
+
+        @retry(stop_max_delay=45 * 60 * 1000,
+               retry_on_result=lambda res: res is False,
+               retry_on_exception=lambda ex: ex is not KeyboardInterrupt)
+        def poll_deploy():
+            log.debug('Fetching deploy status')
+            deployment_result = deploy_poller.result(30)
+            if deploy_poller.done():
+                return deployment_result
+            else:
+                log.info("Waiting for deployment... Current deploy state: {}".format(
+                    deployment_result.properties.provisioning_state))
+                return False
+
+        try:
+            deployment_result = poll_deploy()
+        except:
+            if rollback:
+                azure_wrapper.rmc.resource_groups.delete(group_name)
+            log.exception('Polling raise an exception!')
+            raise Exception('TemplateDeployment failed!')
+
+        outputs = {}
+        # Remove worthless value tags from outputs
+        for k, v in deployment_result.properties.outputs.items():
+            outputs[k] = v['value']
+        new_deploy = cls(group_name, azure_wrapper)
+        new_deploy.outputs = outputs
```

Comment:
if outputs should be assigned like this, shouldn't it be something that the class itself gets / builds? It's odd this is setting a member right after constructing it.